org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/preferences/PersistentModel.java

```diff
@@ -11,37 +11,35 @@
 import java.util.Collection;
 import java.util.List;
 
-import com.google.common.base.Optional;
-
 import org.eclipse.core.resources.IFolder;
-import org.eclipse.core.resources.IResource;
 import org.eclipse.core.runtime.IPath;
 import org.eclipse.jdt.core.IClasspathEntry;
 
 /**
  * Interface to load and store Gradle model elements stored in the workspace plugin state area.
  * <p/>
- * The {@link Optional} return type on the getters mark whether a value is present in the model.
+ * Each getter may return {@code null} values if the model element is not available. If a
+ * {@code null} value is passed to a setter then the model element is deleted.
  *
  * @author Donat Csikos
  */
 public interface PersistentModel {
 
-    Optional<IPath> getBuildDir();
+    IPath getBuildDir();
 
-    void setBuildDir(Optional<IPath> buildDirPath);
+    void setBuildDir(IPath buildDir);
 
     Collection<IPath> getSubprojectPaths();
 
     void setSubprojectPaths(Collection<IPath> subprojectPaths);
 
-    Optional<List<IClasspathEntry>> getClasspath();
+    List<IClasspathEntry> getClasspath();
 
     void setClasspath(List<IClasspathEntry> classpath);
 
-    Collection<IResource> getDerivedResources();
+    Collection<IPath> getDerivedResources();
```

Comment:
Does it make sense for these lists to be nullable? I don't see a semantical difference between an empty or absent list of derived resources. That would get rid of the null checks further down.

If we only saved the model at the end of synchronization, then we could guarantee that:

- either the whole model is there with valid information
- or the whole model is absent

Then we don't need individual null checks and don't need to allow null as a value.